### PR TITLE
Fix the misuse of hostname and address in gp_segment_configuration

### PR DESCRIPTION
**Pull request by gfphoenix78**
gp_segment_configuration has two confusing columns for
addressing the segment: `hostname` and `address`. We've
discussed their semantics in the mail-list: 
https://groups.google.com/a/greenplum.org/g/gpdb-dev/c/LfusrgthupY

In short, `address` is used internally and it's usually a private
IP address or name. `hostname` may be publicly accessible
and the utility tool uses it to run some operations per-host,
e.g. install exactly one copy of a binary package per-host.

WAL replication is internal for GPDB cluster, it should absolutely
use the `address` instead of `hostname` in gp_segment_configuration.
The address for WAL replication is saved in the `primary_conninfo`
in `recovery.conf` on the data directory of the mirrors.

This commit fixes the utility tools to use the `address` and adds
test cases to ensure that the address for WAL replication is the
desired internal address, rather than the hostname. The affected
utility tools include:
gpinitsystem, gpinitstandby, gprecoverseg, gpexpand, gpaddmirrors.

Here are the details:
[demo-cluster]: Use the unicast IP address as the `address` column in
                `gp_segment_configuration`, which is different to the
                `hostname` field for all segments, so we know whether
                it gets from `address` or `hostname`.
                
[gpinitsystem]: Fix a bug(issue #12218) in gpinitsystem that address
                could not be an IP address in configuration file.
                Make the data directory if it doesn't exist.
                
[gpinitstandby]: It accepts only the standby hostname, so the `address`
                 and `hostname` in gp_segment_configuration are the
                 same values. This commit accepts the standby address
                 and resolves the hostname by the passed address. It's
                 fine for the standby address to be anyone of an IP
                 address, resolvable name or hostname.
                 
[gprecoverseg]: gprecoverseg contains mainly 2 types of recovery:
                1) incremental recovery, uses pg_rewind
                2) full recovery, uses pg_basebackup
                Both of pg_rewind and pg_basebackup may write the recovery.conf,
                The address for WAL replication is the `address` of
                the primary in gp_segment_configuratioin, which is part of
                the conninfo passed to pg_rewind/pg_basebackup.
                
[gpexpand]: gpexpand uses pg_basebackup to create the mirrors.

[gpaddmirrors]: It accepts the address of the mirrors and resolves
                the hostnames by the passed address
                

This PR fixes the following issues:
  * #11090
  * #9060
  * #11092
  * #9531
  * #9132
  * #12218

- [ ] Add tests for the change
- [ ] Document changes
- [ ] Communicate in the mailing list if needed
- [ ] Pass `make installcheck`
- [ ] Review a PR in return to support the community

### CODE REVIEW

**Comment by ashwinstar** [comment]
Lets state in commit that this PR fixes:
- https://github.com/greenplum-db/gpdb/issues/11090
- https://github.com/greenplum-db/gpdb/issues/9060


Plus, just for reference (I know this PR is not targeting to fix) these issues but are relevant. These can be fixed as follow-up.
- https://github.com/greenplum-db/gpdb/issues/11092
- https://github.com/greenplum-db/gpdb/issues/9531
- https://github.com/greenplum-db/gpdb/issues/9132

**Comment by soumyadeep2007** [comment]
> gp_segment_configuration has two confusing columns for
addressing the segment: hostname and address. We've
discussed their semantics in the mail-list:
https://groups.google.com/a/greenplum.org/g/gpdb-dev/c/LfusrgthupY

That thread is very helpful. Thanks for sharing. And yes, `address` is
very confusing, renaming it to `segmentHost` or even `hostname_if` is
probably in our best interest, to convey:
1) It pertains to the segment and can be an interface level hostname/address
for multi-homed setups which is typical for GP.
2) It can be both hostname/IP address.
This is for future versions of course.

> In short, address is used internally and it's usually a private
IP address or name. hostname may be publicly accessible
and the utility tool uses it to run some operations per-host,
e.g. install exactly one copy of a binary package.

Yes, `hostname` may be publicly accessible. We want gprecoverseg traffic
to traverse the interconnect, which is only possible if it uses the per-segment
"address" field.

> To run testcases for this commit, the hostname and address in
gp_segment_configuration must not be the same, as the values
generated by demo cluster.

Yes we would need a multi-host test setup with behave tests. I think this
[PR](https://github.com/greenplum-db/gpdb/pull/11968) uses a multi-host
setup.

**Comment by gfphoenix78** [comment]
> Lets state in commit that this PR fixes:
> 
> * [Recovering operation of gprecoverseg should refers to pg_catalog.gp_segment_configuration.address rather than the server hosname #11090](https://github.com/greenplum-db/gpdb/issues/11090)
> * [replication traffic going to hostname vs address specified in gpinit config #9060](https://github.com/greenplum-db/gpdb/issues/9060)
> 
> Plus, just for reference (I know this PR is not targeting to fix) these issues but are relevant. These can be fixed as follow-up.
> 
> * [gpinitstandby joins standby master server with wrong hostname #11092](https://github.com/greenplum-db/gpdb/issues/11092)
> * [gpinitstandby populates standby master hostname with value passed via -s #9531](https://github.com/greenplum-db/gpdb/issues/9531)
> * [gpinitsystem populates mirror hostname with address value #9132](https://github.com/greenplum-db/gpdb/issues/9132)

@ashwinstar Thank you for collecting these issues.

**Comment by kalensk** [comment]
Can we add this to master first and once settled consider back porting to 6X_STABLE, especially in light of the stability efforts and concerns for regressions, etc.

**Comment by gfphoenix78** [comment]
> Can we add this to master first and once settled consider back porting to 6X_STABLE, especially in light of the stability efforts and concerns for regressions, etc.?

Sure, the PR for master branch is coming. Since the fix for 6X is a little urgent, I did it first for GPDB 6X.

File: gpMgmt/bin/gpinitstandby [hunk 1]
```diff
@@ -161,27 +166,39 @@ def parseargs():
...
+        cache = GpInterfaceToHostNameCache(pool, [options.standby_address], [None])
+        hostname = cache.getHostName(options.standby_address)
+        if not hostname:
+            raise Exception("can't resolve address to hostname: '%s' => '%s'" %
+                            (options.standby_address, options.standby_hostname))
+
+        options.standby_hostname = hostname
+    finally:
+        pool.haltWork()
+
```
[the remark this answers is not in the document]
Comment (gfphoenix78) [reply gpMgmt/bin/gpinitstandby]: Good point. Used `unix.Hostname` to get the fresh hostname and ensure that the standby is alive.

**Comment by ashwinstar** [comment]
This PR is against 6X_STABLE,  we need to bake the solution for master branch first and then revisit for STABLE branch. Hence, closing this PR for now.

**Comment by kalensk** [comment]
Hopefully, this is being tracked and appropriately prioritized for 7X as based on my experience fairly important to avoid "kicking the can down the road."

**Pull request closed without merging**